src/hash.rs

```diff
@@ -0,0 +1,564 @@
+        self.0.as_bytes().cmp(other.0.as_bytes())
+    }
+}
+
+impl fmt::Display for Hash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // result will be 52 bytes
+        let mut res = [b'b'; 52];
+        // write the encoded bytes
+        data_encoding::BASE32_NOPAD.encode_mut(self.as_bytes(), &mut res);
+        // convert to string, this is guaranteed to succeed
+        let t = std::str::from_utf8_mut(res.as_mut()).unwrap();
+        // hack since data_encoding doesn't have BASE32LOWER_NOPAD as a const
+        t.make_ascii_lowercase();
+        // write the str, no allocations
+        f.write_str(t)
+    }
+}
+
+impl FromStr for Hash {
+    type Err = HexOrBase32ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        parse_array_hex_or_base32(s).map(Hash::from)
+    }
+}
+
+impl Serialize for Hash {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            serializer.serialize_str(self.to_string().as_str())
+        } else {
+            self.0.as_bytes().serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Hash {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            let s = String::deserialize(deserializer)?;
+            s.parse().map_err(de::Error::custom)
+        } else {
+            let data: [u8; 32] = Deserialize::deserialize(deserializer)?;
+            Ok(Self(blake3::Hash::from(data)))
+        }
+    }
+}
+
+impl MaxSize for Hash {
+    const POSTCARD_MAX_SIZE: usize = 32;
+}
+
+/// A format identifier
+#[derive(
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+    Default,
+    Debug,
+    MaxSize,
+    Hash,
+    derive_more::Display,
+)]
+pub enum BlobFormat {
+    /// Raw blob
+    #[default]
+    Raw,
+    /// A sequence of BLAKE3 hashes
+    HashSeq,
+}
+
+impl From<BlobFormat> for u64 {
+    fn from(value: BlobFormat) -> Self {
+        match value {
+            BlobFormat::Raw => 0,
+            BlobFormat::HashSeq => 1,
+        }
+    }
+}
+
+impl BlobFormat {
+    /// Is raw format
+    pub const fn is_raw(&self) -> bool {
+        matches!(self, BlobFormat::Raw)
+    }
+
+    /// Is hash seq format
+    pub const fn is_hash_seq(&self) -> bool {
+        matches!(self, BlobFormat::HashSeq)
+    }
+}
+
+/// A hash and format pair
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, MaxSize, Hash)]
+pub struct HashAndFormat {
+    /// The hash
+    pub hash: Hash,
+    /// The format
+    pub format: BlobFormat,
+}
+
+#[cfg(feature = "redb")]
+mod redb_support {
+    use postcard::experimental::max_size::MaxSize;
+    use redb::{Key as RedbKey, Value as RedbValue};
+
+    use super::{Hash, HashAndFormat};
+
+    impl RedbValue for Hash {
```

Comment:
Ah, right. This is one reason to not just reexport blake3::Hash. But we could wrap Hash in a newtype just for the store.